### PR TITLE
feat: Implement device authentication flow

### PR DESCRIPTION
**Pull request by kartikCypherock**
Implements device authentication flow - https://app.clickup.com/t/9002019994/PRF-4742

### CODE REVIEW

**Comment by rohanagarwal94** [comment]
Task linked: [PRF-4742 Device authentication - flow implementation](https://app.clickup.com/t/9002019994/PRF-4742)

**Review by ujjwal-cyph — CHANGES_REQUESTED** [review]
needs refactor as https://github.com/Cypherock/x1_wallet_firmware/pull/255 is merged.

File: apps/manager_app/manager_app.c [hunk 1]
```diff
@@ -102,6 +102,8 @@ void manager_app_main(usb_event_t usb_evt) {
     return;
   }
 
+  LOG_SWV("Query:%d\n", query.which_request);
```
Comment (ujjwal-cyph) [inline apps/manager_app/manager_app.c:105]: for easier tracing

```suggestion
  LOG_SWV("%s (%d) - Query:%d\n", __func__, __LINE__, query.which_request);
```

File: apps/manager_app/manager_app.c [hunk 2]
```diff
@@ -129,5 +132,8 @@ void manager_app_main(usb_event_t usb_evt) {
     }
   }
 
+  // TODO: Check if on-boarding default screen is to be rendered
+  onboarding_set_static_screen();
```
Comment (ujjwal-cyph) [inline apps/manager_app/manager_app.c:136]: is this the right place to reset flow status and device state as follows?
```
  core_status_set_flow_status(0);
  core_status_set_idle_state(CORE_DEVICE_IDLE_STATE_DEVICE_IDLE_STATE_IDLE);
```

**Comment by kartikCypherock** [comment]
> needs refactor as #255 is merged.

Rebased and force pushed

File: apps/manager_app/manager_app.c (hunk 2 above)
Comment (ujjwal-cyph) [reply apps/manager_app/manager_app.c:136]: there refers to the application (here `manager_app`), right?

Comment (kartikCypherock) [reply apps/manager_app/manager_app.c:144]: Added both at appropriate places

File: apps/manager_app/device_authentication.c [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+#include "ui_delay.h"
+
+/*****************************************************************************
+ * EXTERN VARIABLES
+ *****************************************************************************/
+
+/*****************************************************************************
+ * PRIVATE MACROS AND DEFINES
+ *****************************************************************************/
+#define DEVICE_AUTH_RESPONSE_SIZE (MANAGER_AUTH_DEVICE_RESPONSE_SIZE + 20)
```
Comment (ujjwal-cyph) [inline apps/manager_app/device_authentication.c:77]: Lets add a todo for visibility. when all option files are finalized, we will get the final size of the `MANAGER_RESPONSE`

File: apps/manager_app/device_authentication.c [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+    } else {
+      // TODO: Handle error scenario in which an unexpected request_type was
+      // detected
+      // TODO: Early flow exit
+    }
+  } else if (RESULT == state) {
+    /* Handle verification status */
+    bool verified = query.auth_device.result.verified;
+    device_auth_handle_response(verified);
+    usb_clear_event();
```
Comment (ujjwal-cyph) [inline apps/manager_app/device_authentication.c:216]: the `device_auth_handle_response` renders delay screens. delay screens pause the execution. if `usb_clear_event` is called after that, the host gets an invalid state if it does not entertain `status` responses. When testing with bare-bone SDK, new commands fail with error as the buffer is blocked till `usb_clear_event`

Comment (ujjwal-cyph) [reply apps/manager_app/device_authentication.c:216]: look at other places too eg: https://github.com/Cypherock/x1_wallet_firmware/pull/254/files#diff-2b0c4f85f60d96db37e84ce447207de7ff219c9087b6cd5578eed2d7f43c180aR203-R205

Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c]: Hi Ujjwal, as per our discussion with Irshad.. We have added a dummy message from device to host to signal flow completion.

Host will poll for the completion message before starting any other transaction.

File: apps/manager_app/device_authentication.c (hunk 3 above)
Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c:79]: Added

File: apps/manager_app/device_authentication.c (hunk 4 above)
Comment (ujjwal-cyph) [reply apps/manager_app/device_authentication.c]: makes sense. That makes the communication consistent.

File: apps/manager_app/device_authentication.c [hunk 5]
```diff
@@ -0,0 +1,274 @@
...
+  delay_scr_init(ui_text_message_device_authenticating, 100);
+
+  /* First state of the device authentication would be SIGN_SERIAL_NUMBER */
+  device_auth_flow_states_e state = SIGN_SERIAL_NUM;
+  send_serial_signature(&state);
+
+  while (FLOW_COMPLETE != state) {
+    evt_status_t event = get_events(EVENT_CONFIG_USB, MAX_INACTIVITY_TIMEOUT);
+    if (true == event.p0_event.flag) {
+      // TODO
```
Comment (ujjwal-cyph) [inline apps/manager_app/device_authentication.c:267]: if it is straight forward, let's add the handling in this PR itself.

File: apps/manager_app/device_authentication.c [hunk 6]
```diff
@@ -0,0 +1,274 @@
...
+  return;
+}
+
+static void device_auth_usb_cb(usb_event_t usb_evt,
+                               device_auth_flow_states_e *state_ptr) {
+  device_auth_flow_states_e state = *state_ptr;
+
+  /* Decode recieved query using protobuf helpers */
+  manager_query_t query = MANAGER_QUERY_INIT_ZERO;
+  if (false == decode_manager_query(usb_evt.p_msg, usb_evt.msg_size, &query)) {
```
Comment (ujjwal-cyph) [inline apps/manager_app/device_authentication.c:190]: in addition to decoding status, add `check_manager_request` too.

Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c:191]: added this check, thanks

File: apps/manager_app/device_authentication.c (hunk 5 above)
Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c]: added for now, but we need to make common p0 handlers soon

File: apps/manager_app/device_authentication.c [hunk 7]
```diff
@@ -0,0 +1,286 @@
...
+
+  if (false == check_manager_request(&query, MANAGER_QUERY_AUTH_DEVICE_TAG)) {
+    // TODO: Handle bad data error
+    // TODO: Early flow exit
+    return;
+  }
+
+  /* Decode recieved request_type */
+  pb_size_t request_type = get_request_type(
+      (const manager_auth_device_request_t *)&(query.auth_device));
```
Comment (ujjwal-cyph) [inline apps/manager_app/device_authentication.c:205]: unnecessary typecast

Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c]: corrected

File: apps/manager_app/device_authentication.c [hunk 8]
```diff
@@ -0,0 +1,367 @@
...
+  return next_state;
+}
+
+/*****************************************************************************
+ * GLOBAL FUNCTIONS
+ *****************************************************************************/
+void device_authentication_flow(const manager_query_t *query) {
+  if (false == check_manager_request(query, MANAGER_QUERY_AUTH_DEVICE_TAG)) {
+    // TODO: Handle bad data error
+    // TODO: Early flow exit
```
Comment (amanCypherock) [inline apps/manager_app/device_authentication.c:279]: For now atleast add a return here.

Comment (amanCypherock) [reply apps/manager_app/device_authentication.c:279]: and in the below condition

File: apps/manager_app/device_authentication.c [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+      *result_valid = true;
+      next_state = SIGN_RANDOM_NUM;
+      break;
+    }
+    case MANAGER_AUTH_DEVICE_REQUEST_CHALLENGE_TAG:
+    case MANAGER_AUTH_DEVICE_REQUEST_RESULT_TAG:
+    default: {
+      usb_clear_event();
+      break;
+    }
```
Comment (amanCypherock) [inline apps/manager_app/device_authentication.c:199]: Set *result_valid = false here;
```suggestion
    default: {
      *result_valid = false;
      usb_clear_event();
      break;
    }
```

File: apps/manager_app/device_authentication.c [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+
+  return next_state;
+}
+
+static device_auth_state_e sign_random_handler(const manager_query_t *query,
+                                               manager_result_t *result,
+                                               bool *result_valid) {
+  pb_size_t request_type = get_request_type(&query->auth_device);
+  device_auth_state_e next_state = SIGN_RANDOM_NUM;
+
```
Comment (amanCypherock) [inline apps/manager_app/device_authentication.c:210]: Can we set `*result_valid = false;` here, so it doesn't rely on what value was assigned to it?

Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c:209]: Hey Aman, I have fixed the logic in flow loop using your PR as reference!

File: apps/manager_app/device_authentication.c (hunk 8 above)
Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c]: I have added the handling as per our discussion today.

File: apps/manager_app/device_authentication.c [hunk 11]
```diff
@@ -0,0 +1,343 @@
...
+      continue;
+    } else {
+      valid_query = true;
+    }
+
+    switch (state) {
+      case SIGN_SERIAL_NUM: {
+        state = sign_serial_handler(query, &result);
+        break;
+      }
```
Comment (amanCypherock) [inline apps/manager_app/device_authentication.c:314]: reject and exit the flow if invalid query is received in this state.

Comment (kartikCypherock) [reply apps/manager_app/device_authentication.c:316]: fixed

**Review by amanCypherock — APPROVED** [review]

**Comment by amanCypherock** [comment]
Before merging this PR, Update the submodule after the PR for it is merged.

**Pull request merged**